Vizzuality/marxan-cloud
# document JSON:API responses for Scenarios + add initial support for pagination

### Overview

Mostly OpenAPI documentation updates (Scenarios) + update shape of Scenario responses as JSON:API

We also add initial support for pagination thanks to an updated version of `BaseService`.

Some of the pagination code is actually within this PR itself, going to be moved to `BaseService` proper once we merge this PR.

In order to be able to iterate more quickly on changes that involve `BaseService`, and also so that we don't have to upstream features/changes that are very specific to this specific API, I have introduced a `AppBaseService` class which inherits from `BaseService` and adds its own stuff where needed. Basically a simple façade pattern to provide some decoupling from the upstream `BaseService`.

Along the way I have unplugged some of the faker code, plugging in instead the actual service code that interacts with the database. Transition to real data is not complete yet.

Since this has been a long-running PR anyway, I ended up sneaking in some documentation and naming changes requested by the frontend team (in `AuthenticationModule`) as well as updating the API base image to the latest NodeJS LTS version and using a specific Alpine release for stricter pinning of the base image.

This PR does not include new e2e tests for Scenarios - I'll add them in a separate PR.

### Designs

N/A

### Testing instructions

- The API image should build 😬 
- OpenAPI documentation for Scenarios should be in JSON:API format

- Try creating ~20 new scenarios, e.g. from variations of this payload:

```
{
    "name": "mollit nulla",
    "type": "marxan",
    "projectId": "b1d3cf45-9412-4ac3-a28f-b4e4f99009b0",
    "description": "dolore mollit",
    "metadata": {},
    "country": "esp",
    "adminRegionId": "b1d3cf45-9412-4ac3-a28f-b4e4f99009b0",
    "extent": {
        "type": "Polygon",
        "coordinates": [
            [
                [-10.0, -10.0],
                [10.0, -10.0],
                [10.0, 10.0],
                [-10.0, -10.0]
            ]
        ]
    },
    "numberOfRuns": 100,
    "boundaryLengthModifier": 0,
    "status": "created"
}
```

- Request a list of scenarios, with (for example) `page[size]=5` and `page[number]=2` - in this case we should get scenarios 6 through to 10 (in whatever db order, since we don't support proper sorting by specific columns just yet).

### Feature relevant tickets

https://www.pivotaltracker.com/story/show/176719309
https://www.pivotaltracker.com/story/show/176680127

---

## Checklist before submitting

      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
